documentation/dataplane/dash-flow-api.md

```diff
@@ -0,0 +1,648 @@
+| -------------------------------------------------- | ------------------------------------------------- | ------------------------------------------------------------ |
+| SAI_SWITCH_ATTR_FLOW_BULK_GET_SESSION_EVENT_NOTIFY | `sai_flow_bulk_get_session_event_notification_fn` | The callback function for receiving events on flow bulk get session event notification. |
+
+```c
+/**
+ * @brief bulk flow get event type
+ */
+typedef enum _sai_flow_bulk_get_session_event_t
+{
+    SAI_FLOW_BULK_GET_SESSION_FINISHED,
+
+    SAI_FLOW_BULK_GET_SESSION_FLOW_ENTRY,
+
+} sai_flow_bulk_get_session_event_t;
+
+/**
+ * @brief Notification data format received from SAI HA set callback
+ *
+ * @count attr[attr_count]
+ */
+typedef struct _sai_flow_bulk_get_session_event_data_t
+{
+    sai_flow_bulk_get_session_event_t event_type;
+
+    sai_object_id_t flow_bulk_session_id;
+  
+    sai_flow_entry_t *flow_entry;
+  
+    uint32_t attr_count; 
+  
+    sai_attribute_t *attr_list; 
+
+} sai_flow_bulk_get_session_event_data_t;
+
+/**
+ * @brief dash flow get bulk session notification
+ *
+ * Passed as a parameter into sai_initialize_switch()
+ *
+ * @count data[count]
+ *
+ * @param[in] count Number of notifications
+ * @param[in] data Array of flow bulk get session events
+ */
+typedef void (*sai_flow_bulk_get_session_event_notification_fn)(
+        _In_ uint32_t count,
+        _In_ const sai_flow_bulk_get_session_event_data_t *flow_bulk_get_session_event_data);
+```
+
+### Protobuf-based flow programming
+
+In addition to the flow state attributes, the flow state can be represented using protobuf for high-efficiency. The attribute of the flow entry is SAI_FLOW_ENTRY_ATTR_FLOW_DATA_PB.
+
+Although the content of both attributes and protobuf may be identical, their applications differ. Attributes enable incremental updates to individual properties, whereas protobuf necessitates a complete update.
+
+```protobuf
+syntax = "proto3";
+
+message MacAddress {
+  bytes address = 1 [(validate.rules).bytes.len = 6];  // MAC address bytes
+}
+
+message IpAddress {
+  enum AddressType {
+    IPV4 = 0;
+    IPV6 = 1;
+  }
+
+  AddressType type = 1;  // IP address type (IPv4 or IPv6)
+  bytes address = 2;  // IP address bytes
+}
+
+message SaiDashFlowKey {
+  uint64 eni_mac = 1;  // ENI MAC address
+  IpAddress src_ip = 2;  // Source IP address
+  IpAddress dst_ip = 3;  // Destination IP address
+  uint8 ip_proto = 4; // IP Protocol
+  uint32 src_port = 5;  // Source port
+  uint32 dst_port = 6;  // Destination port
+}
+
+message SaiDashFlowState {
+  uint32 version = 1;  // SAI_FLOW_ENTRY_ATTR_VERSION
+  uint32 dash_flow_action = 2;  // SAI_FLOW_ENTRY_ATTR_DASH_FLOW_ACTION
+  uint32 meter_class = 3;  // SAI_FLOW_ENTRY_ATTR_METER_CLASS
+  bool is_unidirectional_flow = 4;  // SAI_FLOW_ENTRY_ATTR_IS_UNIDIRECTIONAL_FLOW
+  uint32 underlay_vni = 5;  // SAI_FLOW_ENTRY_ATTR_UNDERLAY_VNI
+  IpAddress underlay_sip = 6;  // Underlay source IP address
```

Comment:
Should we use underlay0 and underlay1 to be consistent with the SAI_FLOW_ENTRY_ATTRs listed earlier in this document?

Reply:
Sure, I have changed these.